repose-aggregator/components/filters/attribute-mapping-policy-validation-filter/src/main/scala/org/openrepose/filters/attributemapping/AttributeMappingPolicyValidationFilter.scala

```diff
@@ -0,0 +1,146 @@
+        chain.doFilter(
+          new HttpServletRequestWrapper(
+            httpServletRequest,
+            normalizedPolicyInputStream),
+          response)
+      case Failure(exception) =>
+        exception match {
+          case uhme: UnsupportedHttpMethodException =>
+            logger.debug("Unsupported HTTP method -- no validation performed", uhme)
+            chain.doFilter(request, response)
+          case ucte: UnsupportedContentTypeException =>
+            logger.debug("Unsupported Content-Type -- validation failed", ucte)
+            httpServletResponse.sendError(
+              SC_UNSUPPORTED_MEDIA_TYPE,
+              ucte.message)
+          case e@(_: SaxonApiException | _: TransformerException | _: JsonProcessingException | _: JsonParseException) =>
+            logger.debug("Validation failed", e)
+            httpServletResponse.sendError(
+              SC_BAD_REQUEST,
+              "Failed to validate attribute mapping policy in request")
+        }
+    }
+  }
+
+  def validateHttpMethod(request: HttpServletRequest): Try[Unit.type] = {
+    if ("PUT".equalsIgnoreCase(request.getMethod)) Success(Unit)
+    else Failure(UnsupportedHttpMethodException(s"${request.getMethod} is not a supported HTTP method"))
+  }
+
+  def getPolicyAsXmlSource(contentType: String, policy: InputStream): Try[Source] = {
+    val contentTypeLowerCase = contentType.toLowerCase
+    val requestStreamSource = new StreamSource(policy)
+
+    if (contentTypeLowerCase.contains("json")) {
+      Try {
+        val outPolicyXml = new XdmDestination
+        AttributeMapper.policy2XML(requestStreamSource, outPolicyXml)
+        outPolicyXml.getXdmNode.asSource
+      }
+    } else if (contentTypeLowerCase.contains("xml")) {
+      Success(requestStreamSource)
+    } else {
+      Failure(UnsupportedContentTypeException(s"$contentType is not a supported Content-Type"))
+    }
+  }
+
+  def getPolicyAsInputStream(contentType: String, policy: Source): Try[ServletInputStream] = {
+    val contentTypeLowerCase = contentType.toLowerCase
+
+    if (contentTypeLowerCase.contains("json")) {
+      Try {
+        // Transform the normalized policy Source into bytes
+        val jsonOutput = new ByteArrayOutputStream()
+        val jsonSerializer = AttributeMapper.processor.newSerializer(jsonOutput)
+        AttributeMapper.policy2JSON(policy, jsonSerializer, validate = false, XSDEngine.AUTO.toString)
+
+        // Wrap the policy bytes with a ServletInputStream
+        new ServletInputStreamWrapper(new ByteArrayInputStream(jsonOutput.toByteArray))
+      }
+    } else if (contentTypeLowerCase.contains("xml")) {
+      Try {
+        // Transform the normalized policy Source into bytes
+        val transformer = transformerFactory.newTransformer()
+        val xmlOutput = new ByteArrayOutputStream()
+        val xmlStreamResult = new StreamResult(xmlOutput)
+        transformer.transform(policy, xmlStreamResult)
```

Comment:
You have to understand the internals of Saxon to know how to avoid using a transformer, but here's how you would do it...

````scala
  AttributeMapper.processor.writeXdmValue (new XdmNode(policy.asInstanceOf[NodeInfo]),  
   AttributeMapper.processor.newSerializer(xmlOutput))
````

You have to know that the source is of type ```NodeInfo``` which is only true if you're using Saxon.  The approach you took will pretty much always work though.